### PR TITLE
[APS-496] Introduce 'assessment allocated' domain event type

### PR DESCRIPTION
**Pull request by danlivings-dxw**
> See [APS-496 on the Approved Premises Jira board](https://dsdmoj.atlassian.net/browse/APS-496).

This PR introduces the `approved-premises.assessment.allocated` domain event type. This is emitted in two scenarios:

1. when an application is submitted and the assessment is first generated; or
2. when the assessment is reallocated.

Additionally, the event contains information about whether the assessment was assigned automatically.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/domainevents/DomainEventDescriber.kt [hunk 1]
```diff
@@ -88,6 +90,15 @@ class DomainEventDescriber(
     return event.describe { "The assessment was appealed and ${it.eventDetails.decision.value}. The reason was: ${it.eventDetails.decisionDetail}" }
   }
 
+  private fun buildAssessmentAllocatedDescription(domainEventSummary: DomainEventSummary): String? {
+    val event = domainEventService.getAssessmentAllocatedEvent(domainEventSummary.id())
+    return event.describe { ev ->
+      val allocatedTo = ev.eventDetails.allocatedTo?.name ?: "an unknown user"
+      val allocatedDesc = ev.eventDetails.allocatedBy?.let { "by ${it.name}" } ?: "automatically"
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/domainevents/DomainEventDescriber.kt:97]: 👍

**Review by pezholio — APPROVED** [review]
LGTM - thanks for picking this up! 👍

File: src/main/resources/static/domain-events-api.yml [hunk 2]
```diff
@@ -1334,6 +1370,48 @@ components:
...
+        - timestamp
+        - eventType
+        - eventDetails
+    AssessmentAllocated:
+      type: object
+      properties:
+        assessmentId:
+          $ref: '#/components/schemas/AssessmentId'
+        assessmentUrl:
+          $ref: '#/components/schemas/AssessmentUrl'
```
Comment (edavey) [inline src/main/resources/static/domain-events-api.yml:1398]: It's good to include the ID and URL of the assessment being allocated, but I think we always need to include the ID and URL of the *application* also in our domain events. 

The application is the entity from which all related entities flow. For example when a Delius integration consumes a domain event it typically uses `applicationUrl` to provide a link back to the CAS1 service. 

See https://github.com/ministryofjustice/hmpps-probation-integration-services/blob/d59771747e6f2263e5f94509f7cbdba1f249993d/projects/approved-premises-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/approvedpremises/EventDetails.kt#L49-L71

File: src/main/resources/static/domain-events-api.yml [hunk 3]
```diff
@@ -1410,7 +1410,6 @@ components:
         - assessmentId
         - assessmentUrl
         - personReference
-        - allocatedBy
```
Comment (edavey) [inline src/main/resources/static/domain-events-api.yml]: Is this removed by mistake?

**Review by edavey — APPROVED** [review]
👍 looks good -- but I think we're missing the essential:

- `applicationId` and
- `applicationUrl`

properties?

File: src/main/resources/static/domain-events-api.yml (hunk 3 above)
Comment (danlivings-dxw) [reply src/main/resources/static/domain-events-api.yml]: I'd say that an allocation that happens automatically isn't "by" anyone, so would best be represented by `null`.

File: src/main/resources/static/domain-events-api.yml (hunk 2 above)
Comment (danlivings-dxw) [reply src/main/resources/static/domain-events-api.yml:1398]: This has been done.

**Pull request merged**